### PR TITLE
Add custom contract types via plugin config

### PR DESCRIPTION
**Pull request by qduk**
This adds the ability to provide a list of contract types in your plugin configuration for use in the plugin. This is related to issue #103. I have a similar use case in my current work where `Hardware` and `Software` aren't specific enough and the use of a `services` and/or `software maintenance` are needed. Let me know what you guys think! Thanks!

### CODE REVIEW

**Comment by progala** [comment]
Thanks @qduk, I like this idea. I'll have a look at this today and we should merge soon.

**Comment by progala** [comment]
@qduk I noticed that these new contract types don't show up in the filter drop down menu.

**Comment by qduk** [comment]
@progala hm let me look into it tomorrow and update here.

File: development/nautobot_config.py [hunk 1]
```diff
@@ -262,6 +262,7 @@
         "barchart_bar_width": float(os.environ.get("BARCHART_BAR_WIDTH", 0.1)),
         "barchart_width": int(os.environ.get("BARCHART_WIDTH", 12)),
         "barchart_height": int(os.environ.get("BARCHART_HEIGHT", 5)),
+        "contract_types": ["Hardware Maintenance", "Subscription", "Software License", "Services", "Lease"],
```
Comment (bradh11) [inline development/nautobot_config.py:265]: Propose being more clear here that there are additional contract types that are appended to the defaults `Hardware` & `Software`

```suggestion
        "additional_contract_types": ["Hardware Maintenance", "Subscription", "Software License", "Services", "Lease"],
```

File: nautobot_device_lifecycle_mgmt/utils.py [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
     return Coalesce(subquery, 0)
+
+
+def add_custom_contract_types(choices):
+    """Add custom defined choices to existing Contract Type choices.
+
+    Args:
+        choices (tuple): Existing Device Lifecycle Contract Type tuple.
+    """
+    defined_contract_types = PLUGIN_SETTINGS.get("contract_types", [])
```
Comment (bradh11) [inline nautobot_device_lifecycle_mgmt/utils.py:22]:
```suggestion
    defined_additional_contract_types = PLUGIN_SETTINGS.get("additional_contract_types", [])
```

File: nautobot_device_lifecycle_mgmt/utils.py [hunk 3]
```diff
@@ -1,10 +1,23 @@
...
+    return tuple((type, type) for type in defined_contract_types) + tuple(choices)
```
Comment (bradh11) [inline nautobot_device_lifecycle_mgmt/utils.py:23]:
```suggestion
    return tuple((type, type) for type in defined_additional_contract_types) + tuple(choices)
```

**Comment by qduk** [comment]
@progala Can you give it another shot? I was able to see the contract types in the dropdown. 
Screenshot 2023-10-31 at 10 35 00 AM

**Comment by qduk** [comment]
@bradh11 Thanks for the suggestions. Do I need to open this toward a different branch? I saw the other one was closed.

**Comment by progala** [comment]
> @bradh11 Thanks for the suggestions. Do I need to open this toward a different branch? I saw the other one was closed.

The other PR was closed as it was targeting `develop`. We tried to retarget it for ltm1.6 but diff was showing too many unrelated changes. Can you try branching off ltm1.6 and reapplying your changes?

**Review by bradh11 — APPROVED** [review]

File: nautobot_device_lifecycle_mgmt/utils.py [hunk 4]
```diff
@@ -1,10 +1,23 @@
...
+
+
+def add_custom_contract_types(choices):
+    """Add custom defined choices to existing Contract Type choices.
+
+    Args:
+        choices (tuple): Existing Device Lifecycle Contract Type tuple.
+    """
+    defined_additional_contract_types = PLUGIN_SETTINGS.get("additional_contract_types", [])
+    return tuple((type, type) for type in defined_additional_contract_types) + tuple(choices)
```
Comment (progala) [inline nautobot_device_lifecycle_mgmt/utils.py:23]:
```suggestion
    return tuple((contract_type, contract_type) for contract_type in defined_additional_contract_types) + tuple(choices)
```

File: development/nautobot_config.py [hunk 5]
```diff
@@ -138,5 +138,6 @@
         "barchart_bar_width": float(os.environ.get("BARCHART_BAR_WIDTH", 0.1)),
         "barchart_width": int(os.environ.get("BARCHART_WIDTH", 12)),
         "barchart_height": int(os.environ.get("BARCHART_HEIGHT", 5)),
+        "additional_contract_types": ["Hardware Maintenance", "Subscription", "Software License", "Services", "Lease"],
```
Comment (progala) [inline development/nautobot_config.py:141]:
```suggestion
        "additional_contract_types": os.environ.get(
            "DEVICE_LIFECYCLE_ADD_CONTRACT_TYPES",
            "Hardware Maintenance,Subscription,Software License,Services,Lease",
        ).split(","),
```

**Review by progala — CHANGES_REQUESTED** [review]
Small suggestion. Could you please also add tests for `add_custom_contract_types`

**Comment by qduk** [comment]
@progala Test added here as well. Let me know if these suffice. Thanks!

**Pull request merged**